### PR TITLE
[feature] roles/groups and permissions via Kafka

### PR DESCRIPTION
**Pull request by shipperizer**
@artsalliancemedia/thunderstorm

Enable roles and groups handling via kafka (`ts_event`), permissions sent via `send_ts_event`

* `handle_role_data` and `handle_group_data` are now also kafka events if flag enabled
* permissions can be sent via kafka if flag enabled
* `user_uuid` parsed into the JWT token object

### CODE REVIEW

File: test/test_groups.py [hunk 1]
```diff
@@ -87,3 +88,33 @@ def test_delete_group_association_keeps_going_if_group_association_does_not_exis
...
+            'group_uuid': group_uuid,
+            'complex_uuids': [new_uuid] + complex_uuids[:2]
+        }
+    }
+
+    assert db_session.query(ComplexGroupComplexAssociation).count() == 4
+
+    async with handle_group_data.test_context() as agent:
+        await agent.put(message)
+        # allow 1s to celery worker to do his magic
```
Comment (will-norris) [inline test/test_groups.py]: # allow 1s for faust worker to do his magic

Comment (shipperizer) [reply test/test_groups.py]: it is actually the celery worker that deals with the things under the hood

File: thunderstorm_auth/roles.py [hunk 2]
```diff
@@ -144,36 +163,56 @@ def remove_role_orphan_permission_associations(role_uuid, permissions):
...
+                create_role_if_not_exists_task.si(message['uuid'], message['type']),
+                remove_role_orphan_permission_associations_task.si(message['uuid'], message['permissions']),
+                create_role_permission_associations_if_not_exist_task.si(message['uuid'], message['permissions'])
+            )()
+    elif not isinstance(kafka_app, faust.App):
+        raise ThunderstormAuthError(f'{kafka_app} is not a subclass of faust.App')
+    elif isinstance(kafka_app, faust.App):
+        @kafka_app.ts_event(Event(topic='role.data', schema=RoleSchema))
+        async def handle_role_data(message):
+            # TODO @shipperizer remove this when using ts_task decorator
```
Comment (mcanetti) [inline thunderstorm_auth/roles.py]: what's this?

File: thunderstorm_auth/roles.py [hunk 3]
```diff
@@ -144,36 +163,56 @@ def remove_role_orphan_permission_associations(role_uuid, permissions):
...
+    Returns:
+        Celery.task: if kafka_app is None, returns a celery task for handling roles
+        None: if kafka_app is not None
+    """
+    # TODO @shipperizer: change name on the user service so that this can be standardized
+    if kafka_app is None:
+        @ts_task('role.data', schema=RoleSchema())
+        @statsd.timer('tasks.handle_role_data.time')
+        def handle_role_data(message):
+            chain(
```
Comment (mcanetti) [inline thunderstorm_auth/roles.py]: _handle_role_data?

Comment (mcanetti) [inline thunderstorm_auth/roles.py]: _handle_role_data?

File: thunderstorm_auth/permissions.py [hunk 4]
```diff
@@ -165,77 +165,81 @@ def create_permission_model(base):
...
-                        exchange=MESSAGING_EXCHANGE,
-                        routing_key=PERMISSIONS_NEW,
-                    )
-
-            permission.is_sent = True
-            datastore.commit()
-
-
-    return [handle_sync_permissions]
+       for permission in permissions:
```
Comment (shipperizer) [inline thunderstorm_auth/permissions.py]: `permissions` is undefined here bro, missing a line above where u fetch them

Comment (will-norris) [reply thunderstorm_auth/permissions.py]: It's just a query that you are defining, it's not executed until you loop over it anyway - tested this stuff on the agent service and it all works. To be fair this all needed some serious rework as the kafka stuff was being defined in the celery task which made no sense and the signals were being used incorrectly which is why they failed most of the time. Overall I think this is a cleaner/simpler approach

**Pull request closed without merging**